src/interfaces/mining.h

```diff
@@ -22,6 +25,18 @@ class Mining
     //! If this chain is exclusively used for testing
     virtual bool isTestChain() = 0;
 
+    /**
+     * Check a block is completely valid from start to finish.
+     * Only works on top of our current best block.
+     * Does not check proof-of-work.
+     *
+     * @param[out] state details of why a block failed to validate
+     * @param[in] block the block to validate
+     * @param[in] check_merkle_root call CheckMerkleRoot()
+     * @returns false if any of the checks fail
+     */
+    virtual bool testBlockValidity(BlockValidationState& state, const CBlock& block, bool check_merkle_root = true) = 0;
```

Comment:
In commit "rpc: call TestBlockValidity via miner interface" (d8a3496b5ad27bea4c79ea0344f595cc1b95f0d3)

Not important, but it might be good to move the `state` output parameter last. The reason would be to work with the libmultiprocess code generator in case we want to expose this interface over IPC. The code generator only knows how to parse Cap'n Proto declarations, not C++ code, so when it encounters a Cap'n Proto method declaration with output parameters like:

```capnp
testBlockValidity @1 (block: Data, maxTxFee :Bool) -> (state: ValidationState, result :Bool);
```

it just assumes a corresponding C++ method exists with the listed input parameters first and output parameters last.

(Also I just noticed information about output parameters was changed in the developer notes. Suggestion to put output parameters last used used to be mentioned in the "internal interface" style section, but it looks like e66b321fd1ddfffd9bfc59d407ad8f03490b873c from #25092 turned that advise into a bigger, broader rule, which I'm not sure is a great change..)

Reply:
Thanks, I'll change this if I have to retouch.

Reply:
I made a note to do this along with a PR to add something like `waitTipChanged()` 
https://github.com/bitcoin/bitcoin/pull/30200#issuecomment-2160764070

Reply:
Actually, doing this now since I ran into another (minor) issue: #30335